### PR TITLE
[opt](scanner) Control the degree of parallelism of scanner when only limit involved

### PR DESCRIPTION
**Pull request by zhiqiang-hhhh**
For queries like `select * from tbl limit 100;`, we should limit the parallelism to 1.

We already have strategies in our code, but we can not control its behavior. So add session variables to control them.

Previous pr like #33888 #36535 already has done something similar, this pr integrates them.

### CODE REVIEW

File: gensrc/thrift/PaloInternalService.thrift [hunk 1]
```diff
@@ -334,6 +334,9 @@ struct TQueryOptions {
 
   126: optional i32 runtime_bloom_filter_max_size = 16777216;
 
+  128: optional bool enable_adaptive_pipeline_task_serial_read_on_limit = true;
+  129: optional i32 adaptive_pipeline_task_serial_read_on_limit = 8192;
```
Comment (yiguolei) [inline gensrc/thrift/PaloInternalService.thrift:338]: the value should be 1000 in the past?

File: be/src/pipeline/exec/scan_operator.cpp [hunk 2]
```diff
@@ -1423,6 +1417,31 @@ Status ScanOperatorX<LocalStateType>::init(const TPlanNode& tnode, RuntimeState*
     if (tnode.__isset.topn_filter_source_node_ids) {
         topn_filter_source_node_ids = tnode.topn_filter_source_node_ids;
     }
+
+    // The first branch is kept for compatibility with the old version of the FE
+    if (!state->query_options().__isset.enable_adaptive_pipeline_task_serial_read_on_limit) {
```
Comment (yiguolei) [inline be/src/pipeline/exec/scan_operator.cpp:1422]: 封装一个完整的函数在state 中，不要这样直接读取query options，按道理说query otpions 应该是一个私有的对象

File: be/src/pipeline/exec/scan_operator.cpp [hunk 3]
```diff
@@ -1423,6 +1417,31 @@ Status ScanOperatorX<LocalStateType>::init(const TPlanNode& tnode, RuntimeState*
...
+        if (!tnode.__isset.conjuncts || tnode.conjuncts.empty()) {
+            // Which means the request could be fullfilled in a single segment iterator request.
+            if (tnode.limit > 0 && tnode.limit <= 8192) {
+                _should_run_serial = true;
+            }
+        }
+    } else {
+        if (state->query_options().enable_adaptive_pipeline_task_serial_read_on_limit) {
+            int32_t adaptive_pipeline_task_serial_read_on_limit = 8192;
```
Comment (yiguolei) [inline be/src/pipeline/exec/scan_operator.cpp:1431]: 所有访问thrift 里的字段，都要先判断isset，然后再get 里面的值

File: gensrc/thrift/PaloInternalService.thrift (hunk 1 above)
Comment (zhiqiang-hhhh) [reply gensrc/thrift/PaloInternalService.thrift:338]: 在 be 上默认值是 1000，在 FE 默认值是 10000

File: be/src/pipeline/exec/scan_operator.cpp (hunk 2 above)
Comment (zhiqiang-hhhh) [reply be/src/pipeline/exec/scan_operator.cpp:1422]: 改成直接用前面获取到的 query_options 了。

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 4]
```diff
@@ -1246,13 +1251,11 @@ public void setMaxJoinNumberOfReorder(int maxJoinNumberOfReorder) {
...
-                    + "系统会将这个算子的并发度调整为1，以减少简单查询的扇出",
-                    "When a ScanNode has no filter conditions and the limit value is less than this threshold, "
-                            + "the system will adjust the concurrency of this operator to 1 "
-                            + "to reduce the fan-out of simple queries"})
-    public long limitRowsForSingleInstance = 10000;
+    @VariableMgr.VarAttr(name = LIMIT_ROWS_FOR_SINGLE_INSTANCE, setter = "setLimitRowsForSingleInstance",
+            description = {
+                "请使用 adaptive_pipeline_task_serial_read_on_limit",
+                "Please use adaptive_pipeline_task_serial_read_on_limit"})
+    public long limitRowsForSingleInstance = 8192;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1258]: 保持原来的值，10000就行了，万一有一些极限情况，上线后，直接性能下降了

Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1258]: FE BE上参数不一样，都改成10000？

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run external

**Comment by zhiqiang-hhhh** [comment]
run external

File: fe/fe-core/src/main/java/org/apache/doris/planner/ScanNode.java [hunk 5]
```diff
@@ -739,8 +739,21 @@ public int getScanRangeNum() {
...
+
+        if (ctx != null) {
+            if (ctx.getSessionVariable().enableAdaptivePipelineTaskSerialReadOnLimit) {
+                adaptivePipelineTaskSerialReadOnLimit = ctx.getSessionVariable().adaptivePipelineTaskSerialReadOnLimit;
+            } else {
+                return false;
+            }
+        } else {
+            // No connection context, typically for broker load.
+            adaptivePipelineTaskSerialReadOnLimit = 10000;
```
Comment (HappenLee) [inline fe/fe-core/src/main/java/org/apache/doris/planner/ScanNode.java:752]: already set in line 742？

Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/planner/ScanNode.java]: done

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Pull request merged**